shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,391 @@
+   */
+  public void newline(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.newline")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "go" action.
+   */
+  public void go(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.go")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "search" action.
+   */
+  public void search(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.search")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "send" action.
+   */
+  public void send(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.send")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "done" action.
+   */
+  public void done(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.done")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "next" action.
+   */
+  public void next(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.next")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "previous" action.
+   */
+  public void previous(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.previous")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute an "unspecified" action.
+   */
+  public void unspecifiedAction(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.unspecified")
+    );
+  }
+
+  /**
+   * Sets the {@link TextInputMethodHandler} which receives all events and requests
+   * that are parsed from the underlying platform channel.
+   */
+  public void setTextInputMethodHandler(@Nullable TextInputMethodHandler textInputMethodHandler) {
+    this.textInputMethodHandler = textInputMethodHandler;
+  }
+
+  public interface TextInputMethodHandler {
+    // TODO(mattcarroll): javadoc
+    void show();
+
+    // TODO(mattcarroll): javadoc
+    void hide();
+
+    // TODO(mattcarroll): javadoc
+    void setClient(int textInputClientId, @NonNull Configuration configuration);
+
+    // TODO(mattcarroll): javadoc
+    void setEditingState(@NonNull TextEditState editingState);
+
+    // TODO(mattcarroll): javadoc
+    void clearClient();
+  }
+
+  /**
+   * A text editing configuration.
+   */
+  public static class Configuration {
+    public static Configuration fromJson(@NonNull JSONObject json) throws JSONException {
```

Comment:
Maybe document what the JSONObject should be formatted to look like?

Reply:
I'm of two minds about this.  On the one hand I agree with you that it would be really nice for devs to look at the javadoc and see the format.  On the other hand I'm wondering how many days it will take for that comment to fall out of sync with the code.....what do you think?

Reply:
That's the inherent problem with Javadoc and code comments, unfortunately. Right now this is basically taking an untyped `Object object` and then risking exploding at runtime if it doesn't happen to be the right things. It makes sense to me that the right structure should be documented somewhere. If not here than it makes sense to link to an example file or something similar where you can see the keys.